### PR TITLE
Created MLOps Intro to ML Model Deployment and Management Tutorial

### PR DESCRIPTION
**Pull request by james94**
Created a live code data recipe for data pump for batch scoring
Created python script for cli data pump for batch scoring
Created Intro to ML Model Deployment and Management Tutorial:

- Objective, Prerequisites, 7 tasks, next steps, appendices, etc

### CODE REVIEW

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (orendain) [inline MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:847]: So it looks like there are 3 ways to pump in data:

1. Driverless AI pump as described in this section.
2. The offline CLI with help from this linked script.
3. Using the data pump built into the Platform Studio web app.

My 2 cents:
- The data pump build into Studio page is **useful** because it's easy.  They just drop in a URL and things work.
- The offline script is **useful** because someone can see sort of plainly what the HTTP calls look like.
- The DAI approach is **somewhat confusing**.  The code non-trivial (transforming frames from something named `X` and unclear what the schema looks like).  I get that it's a useful feature, but I wonder if it adds any value to an **MLOps** tutorial.

Just sharing my thoughts - not pushing either way.

Comment (orendain) [inline MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:847]: While the offline script is useful, it would require a user download the dataset, which requires more steps to make sure they get it right.  If we're not providing a how-to for this use-case, would it make sense to leave this option out?

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (orendain) [inline MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:891]: Ooooh, this is a good challenge!  👍  An issue I often see with tutorials online is they are written to be `followed once`, and not necessarily act as a reference for regular use.  Adding this challenge seems like a good way to measure if the tutorial helped a user become self-sufficient when browsing around the product.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (orendain) [inline MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:902]: QQ: How does this section play into the larger tutorial?  That credit card use-case seems almost orthogonal to the current tutorial.  The steps are almost identical (except for having to upload the dataset), except it's very concise and less verbose.

I wonder if moving this to some sort of "Quick Start" tutorial instead of making a long tutorial even longer has any benefit.  Just wondering.

File: MLOps/intro-to-ml-model-deployment-and-management/recipes/data-pumper-called-from-cli.py
Comment (orendain) [inline MLOps/intro-to-ml-model-deployment-and-management/recipes/data-pumper-called-from-cli.py:15]: If someone has access to the CSV, I'm not sure it would do any good to also use `datatable` when they could read the fields directly from the CSV.  The code for this is in one of the Studio app files.

I wrote more thoughts on using (or not) this file in a previous comment.

File: MLOps/intro-to-ml-model-deployment-and-management/recipes/hydraulic-data.py
Comment (orendain) [inline MLOps/intro-to-ml-model-deployment-and-management/recipes/hydraulic-data.py:1]: This (same?) file exists in at least one other place, shared by another one of your PRs.

You may run into maintainability problems having different copies floating around.  It may be beneficial to refer to one of the already existing copies of this script instead of introducing it again.  IIRC, the other one may also have had some minor improvements gathered during the review.

**Review by orendain — CHANGES_REQUESTED** [review]
Admittedly, I skipped over the vast majority of the "concepts" section due to time constraints (plus my feedback wouldn't be as valuable as that from someone more familiar with the goals of that section).

**General feedback not included in the comments below:**
- I would _highly recommend_ running an image optimizer against this branch before merging.  Images of this size will _very_ noticeably slow down the speed of a user's page render.  They'll also make this repository generally slower to work with no easy way to undo it once merged (even if images are optimized later in a future commit).  I use [ImgBot](https://imgbot.net/) for this.
- I felt there wasn't much of an "Aha!" moment when deploying models.  It's my understanding that deployment is generally very hard for most companies - but with the click of a single button in MLOps it's all taken care of.  Maybe some sort of declaration that the user clicking "Deploy" is all it takes would help create this moment.  I don't know - just thinking aloud.
- I appreciate all of the media/screenshot!  While I didn't personally follow the tutorial images, I imagine this takes the guess work from trying to find where on the screen things are.


**Just some thoughts ... 🤔:**
- Because there are so many screenshots (yay!), many of them look very similar, if not identical.  Some screenshots are the same as the one before it, with only a drop-down open.  I wonder if consolidating several of these screenshots into gifs would not only shorten the tutorial, but be more clear than static shots.
- It was a bit hard to make out when certain strings were referring to sections of a screenshot, since there was no bolding, tabling, etc.  I'm guessing this was just to get a draft out, so apologies if this was intended! 🙏

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:448]: @orendain yes this link is supposed to take the user to the someplace. The source is a video from AI Foundations course. I show timeframe, so user can keep it in mind for when they look at the video.

Comment (orendain) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:448]: To clarify, this comment is about the URL and not the linked text.  The URLs are currently `https://training.h2o.ai/products/module-7-machine-learning-operations?packages%5B%5D=47590&in_package=47590&sort_by=package_order&ref=package&ref_id=47590#tab-product_tab_contents__4`, while the shorter version `https://training.h2o.ai/products/module-7-machine-learning-operations4` seems to lead to the same place.  So the longer URL might not be necessary.

File: MLOps/intro-to-ml-model-deployment-and-management/recipes/data-pumper-called-from-cli.py
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/recipes/data-pumper-called-from-cli.py:15]: @orendain I referenced the studio app code for the live code py example. This script for cli could be replaced with the one from the studio app with slight modification.

**Comment by james94** [comment]
@orendain I used an image optimizer called https://github.com/VirtuBox/img-optimize, but it seems like the image's size did not decrease. Can you show me how to use imgbot? When does it come into play?

**Comment by james94** [comment]
@orendain I feel like the aha! moment should come after the user reads through the concepts section and they learn about how difficult the process of doing model deployment is and then they learn how to do it with a few clicks in MLOps. Maybe I need to make that connection briefly of the challenges that were discussed in the concepts and remind them how we just deployed a model, which was a very challenging task.

**Comment by james94** [comment]
@orendain I do agree with you on adding gifs as they can make it easier for the user to follow along. You are correct, submitting this PR was the start to getting a draft out, so that there is more attention put toward this tutorial and so we are closer to completing the process of publishing it. Yes my intention is to add some gifs where needed, crop some images, add green rectangle shapes in some images and potentially tables too.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:24]: @orendain I changed the filename to be lowercase. Good point on keeping it consistent with the rest of the images.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:48]: Okay thanks for the heads up. I usually put the version numbers at least in 1 place of the tutorial. These numbers are based on the version numbers used in Test Drive. @Castro38 should we include the version number in this tutorial?

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:8]: @orendain thanks for the suggestion. I replaced ampersand with "and", it should work in the new commit.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:11]: @orendain thank you for pointing out this link issue. I made the change. It will be in the new commit. I will check if it works properly.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:448]: @orendain thanks, I will make replace the long URL with the shorter URL. That update will be in the new commit.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:461]: @orendain there is not really a benefit to pointing out the logo and header. I can remove it.

**Comment by orendain** [comment]
> @orendain I used an image optimizer called https://github.com/VirtuBox/img-optimize, but it seems like the image's size did not decrease. Can you show me how to use imgbot? When does it come into play?

I'm not sure if there's a setting for getting imgbot to run on branches, but the way I usually run it is admittedly kind of hacky:
- Fork the repo.
- Make my proposed change and push it to `master` (or whatever the default branch is) (of my own forked repo).
- ImgBot automatically creates a PR against `master` with the optimized images.
- Then I use that branch or just copy/paste those images.

I'm surprised that the other repo didn't reduce the size of the images.  Maybe another optimizer would have different results, or maybe something else is at play (perhaps dimensions of the image [I haven't checked]).


> I feel like the "aha!" moment should come after the user reads through the concepts section and they learn about how difficult the process of doing model deployment is and then they learn how to do it with a few clicks in MLOps. Maybe I need to make that connection briefly of the challenges that were discussed in the concepts and remind them how we just deployed a model, which was a very challenging task.

Ah, understood!  Thanks!


> @orendain I do agree with you on adding gifs as they can make it easier for the user to follow along. You are correct, submitting this PR was the start to getting a draft out, so that there is more attention put toward this tutorial and so we are closer to completing the process of publishing it. Yes my intention is to add some gifs where needed, crop some images, add green rectangle shapes in some images and potentially tables too.

Sounds good!  IMHO static pictures are absolutely fine for a first iteration 👍 .

File: MLOps/intro-to-ml-model-deployment-and-management/recipes/hydraulic-data.py
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/recipes/hydraulic-data.py:1]: @orendain I will reference the hydraulic-data.py data recipe from the recipes repo.

File: MLOps/intro-to-ml-model-deployment-and-management/recipes/data-pumper-called-from-cli.py
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/recipes/data-pumper-called-from-cli.py:15]: @orendain I will remove this py script for now.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:902]: @orendain I was debating whether to keep this appendix section or not. I was leaning toward removing it. I can commented it out for now in case the team later wants to leverage it for an MLOps quickstart tutorial.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:891]: @orendain thanks. It would be good to be able to measure if the tutorial helped users become self-sufficient.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:847]: @orendain I agree with you. So, I removed the comment on using an alternative python script.

Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:847]: @orendain I do remember the studio app being very user friendly, but with the current version of MLOps test drive that I have, the data pump only supports the credit card dataset. So, I created a live code data recipe to act as a data pump. It was sort of a hack since I don't believe data recipe was meant to do that. I decided not to use python from cli since the user would be pumping data outside the environment of DAI and MLOps. It would be nice to have live code python editor in MLOps like DAI.

Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:847]: @orendain for now I will keep the DAI approach to pump data to MLOps. But I would prefer to use the studio app or if MLOps had a built in live code python editor, then that would also be a good potential solution.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:811]: @orendain I made sure to add extra info on where the user can find the `scoring_endpoint`. I have them refer to the beginning of task 5 where we look at a figure that shows dropdown menu.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:761]: @orendain I made the update

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:752]: @orendain I made the update

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:713]: @orendain I made the update

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:536]: @orendain thank you for the suggestion. I made the update, hopefully it is easier for the reader to notice that `refer to...` statement now.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:491]: @orendain thanks for pointing this out. For each figure, I made sure to bold certain text when appropriate and I also added green boxes and bubbled numbers or letters when appropriate in the image. I hope this helps users follow along better.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:336]: @orendain I did split this paragraph into 2 separate paragraphs.

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (james94) [reply MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:282]: @orendain I did separate this paragraph into 2 separate paragraphs.

**Comment by james94** [comment]
> I would highly recommend running an image optimizer against this branch before merging. Images of this size will very noticeably slow down the speed of a user's page render. They'll also make this repository generally slower to work with no easy way to undo it once merged (even if images are optimized later in a future commit). I use ImgBot for this.

I need to create a new fork with this particular branch. Then I can try to run ImgBot to optimize the images for this PR.

> I felt there wasn't much of an "Aha!" moment when deploying models. It's my understanding that deployment is generally very hard for most companies - but with the click of a single button in MLOps it's all taken care of. Maybe some sort of declaration that the user clicking "Deploy" is all it takes would help create this moment. I don't know - just thinking aloud.

I did add a statement that reminds the user about the concepts section on how hard it is deploy a ML model and that they did it in 3 clicks.

I will need to incorporate some gifs for the next iteration. For now, I am going to commit the latest updates and then work on forking from the latest repo and see if I can also fork this branch.

**Comment by james94** [comment]
@orendain when you get a chance, can you go through the latest commit? ecce3cb

**Comment by james94** [comment]
@orendain I merged in my PR from forked repo master to this branch. Check commit aef7061

- Created a forked repo with forked branch of MLOps tutorial
- Then had ImgBot auto run to optimize images in forked repo for MLOps
- Added GIFs in the tutorial where I believed appropriate

There is one concern I have not related to the MLOps tutorial. I did pull in the latest updates from the parent master repo, so my forked master branch would be up to date.

~~~bash
git remote add upstream git://github.com/h2oai/tutorials.git
git fetch upstream
git pull upstream master
~~~

File: MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md
Comment (orendain) [inline MLOps/intro-to-ml-model-deployment-and-management/intro-to-ml-model-deployment-and-management.md:16]: 👍 for removing the appendix section that was commented! 👍   There are pros to not having commented code/content part of the checked-in code.  Very clean!

**Review by orendain — APPROVED** [review]
I like the changes!  Readability and ease of following was definitely improved for me! 👍 

Just a couple of notes:
- I like how there were multiple screenshots that were combined into single gifs.  It shortened the tutorial _and_ made it easier to follow as I was no longer trying to connect the strings between sequential screenshots - but instead had to follow the one gif.
- The gifs seem quite fast for me.  Not sure if I'm seeing them at the right speed, but they appear sped up on my end.
- I would _highly recommend_ that the optimized images _not_ relevant to this tutorial be moved to another PR.  It would keep the scope of the PR to just the new tutorial.  Modifying other tutorials you might also need to double check with the other authors, or at least give them a heads up - and moving the other images into some sort of "Optimize all images" PR may make it easier to share/review/explain awesome new process/etc.  I'd defer to what @Castro38 thinks is best on this.

**Comment by james94** [comment]
@orendain 

> I would highly recommend that the optimized images not relevant to this tutorial be moved to another PR. It would keep the scope of the PR to just the new tutorial. Modifying other tutorials you might also need to double check with the other authors, or at least give them a heads up - and moving the other images into some sort of "Optimize all images" PR may make it easier to share/review/explain awesome new process/etc. I'd defer to what @Castro38 thinks is best on this.

I am going to undo pull request from forked repo branch to parent repo branch. Then update the content on this parent repo branch and images and gifs. So, only optimized images for this tutorial will be in this PR.

**Pull request closed without merging**
